File: 프로그래머스/3/43162. 네트워크/네트워크.cs
using System;

public class Solution {
    int result=0;
    bool[] isSelected;
    int N = 0;
    public int solution(int n, int[,] computers) {
        N = n;
        isSelected = new bool[N];
        
        for(int i=0;i<N;i++){
            if(isSelected[i]) continue;
            result++;
            isSelected[i] = true;
            dfs(i, computers);
        }
        
        return result;
    }
    
    void dfs(int v, int[,] status){
        //지금 내가 들어온 정점에 대하여 방문 체크 진행
        for(int i=0;i<N;i++){
            if(i==v) continue;
            //갈 수 없는 곳이면 못간다.
            if(status[v,i]==0) continue;
            
            //이미 가본 곳이면 안간다.
            if(isSelected[i]) continue;
            
            isSelected[i]=true;
            dfs(i, status);
        }
        
        
    }
}

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;
class Solution {
   
    int[] parents;
    boolean[] isSelected;
    public int solution(int n, int[][] computers) {
        parents = new int[n+1];
        isSelected = new boolean[n+1];
        
        // 부모값 초기화
        for(int i=0;i<n+1;i++){
            parents[i] = i;
        }
        
        for(int i=0;i<n;i++){
            for(int j=0;j<n;j++){
                if(i== j) continue;
                if(computers[i][j]==0) continue;
                
                union(i+1, j+1);
            }
        }
        
        System.out.println(Arrays.toString(parents));
        
        int cnt =0;
        for(int i=1;i<n+1;i++){
            int p = find(i);
            if(isSelected[p]) continue;
            isSelected[p] = true;
            cnt++;
        }
        return cnt;
        
    }
    
    public void union(int x, int y){
        x = find(x);
        y = find(y);
        
        if(x == y) return;

        if(x < y){
            parents[y] = x;
        } else {
            parents[x] = y;
        }
    }
    
    public int find(int x){
        if(parents[x] == x){
            return x;
        }
        
        return find(parents[x]);
    }
    
}